cooked-validators/src/Cooked/Tx/Constraints.hs

```diff
@@ -53,11 +53,11 @@ class ToLedgerConstraint constraint where
   toLedgerConstraint :: constraint -> LedgerConstraint a
 
 instance ToLedgerConstraint MiscConstraint where
-  extractDatumStr (SpendsScript _validator _redeemer (_out, datum)) =
+  extractDatumStr (SpendsScript _ _ (_, Pl.ScriptChainIndexTxOut _ _ (Right datum) _)) =
     M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData datum) (show datum)
   extractDatumStr _ = M.empty
```

Comment:
Shall this return an error instead? That is, unless there is some legitimate use of this function when the datum is missing.

Reply:
This function is there to populate a `Map DatumHash String` for pretty printing. A constraint whose `ChainIndexTxOut` is a `Left datumHash` instead of a `Right datum` already fails at runtime during validation with a plutus app error `DatumNotFound` that is pretty clear.

All the constraints which deal with no datum (including `SignedBy` or `Before` which have nothing to do with datum) return an empty map. I thought it was more consistent to return an empty map in case a `SpendsScript` constraint has no datum to expose.

But there is more:
I think this function is useless for anything else than `PaysScript` and `PaysPKWithDatum`. When we reach a `SpendsScript`, the datum is already known from a `PaysScript`, there is no need to add it to the map used to pretty print the `UtxoState`. In case there is no associated `PaysScript` with the right datum, the transaction would fail anyway so there would be no `UtxoState` to print.
Replacing the whole implementation by `M.empty` should not even change anything.

Maybe this deserves a dedicated issue and/or PR.